control/grpc.py

```diff
@@ -871,12 +872,35 @@ def set_ana_state_safe(self, ana_info: pb2.ana_info, context=None):
                 return pb2.req_status(status=ret_recycle , error_message=errmsg)
         return pb2.req_status(status=True)
 
+    def choose_anagrpid_for_namespace(self, nsid) ->int:
+        grps_list = self.ceph_utils.get_number_created_gateways(self.gateway_pool, self.gateway_group)
+        for ana_grp in grps_list:
+            if not self.clusters[ana_grp]: # still no namespaces in this ana-group - probably the new GW  added
+                self.logger.info(f"New GW created: chosen ana group {ana_grp} for ns {nsid} ")
+                return ana_grp
+        #not found ana_grp .To calulate it.  Find minimum loaded ana_grp cluster
+        ana_load = {}
+        min_load = 2000
+        chosen_ana_group = 0
+        for ana_grp in self.clusters:
+            ana_load[ana_grp] = 0;
+            for name in self.clusters[ana_grp]:
+                ana_load[ana_grp] += self.clusters[ana_grp][name] # accumulate the total load per ana group for all ana_grp clusters
+        for ana_grp in ana_load :
+            self.logger.info(f" ana group {ana_grp} load =  {ana_load[ana_grp]}  ")
+            if ana_load[ana_grp] <=  min_load:
+                min_load = ana_load[ana_grp]
+                chosen_ana_group = ana_grp
+                self.logger.info(f" ana group {ana_grp} load =  {ana_load[ana_grp]} set as min {min_load} ")
+        self.logger.info(f"Found min loaded cluster: chosen ana group {chosen_ana_group} for ns {nsid} ")
+        return chosen_ana_group
+
     def namespace_add_safe(self, request, context):
         """Adds a namespace to a subsystem."""
 
         peer_msg = self.get_peer_message(context)
         nsid_msg = self.get_ns_id_message(request.nsid, request.uuid)
-        self.logger.info(f"Received request to add a namespace {nsid_msg}to {request.subsystem_nqn}, context: {context}{peer_msg}")
+        self.logger.info(f"Received request to add a namespace {nsid_msg}to {request.subsystem_nqn},  ana group {request.anagrpid}  context: {context}")
```

Comment:
You have a merge issue here. Please add "{peer_msg}" after "{context}".

Reply:
done